### ISSUE TITLE
[lib] readonly modifier is missing in a lot of places

### ISSUE THREAD

**Issue by Conaclos** [microsoft/TypeScript#47823]
# lib Update Request

I have code that does not type-check because of missing readonly modifier.

For instance the following code does not type-check

```ts
declare const vals: readonly number[]
Math.max.apply(null, vals)
```
[Playground link](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gNyhNAXPHFMKhAJ7woCuAtgEYgwDaAugLABQAslBgAsAdHSgAPIVAAOUygApaECABoceNAEogA)

This is caused by the following signatures:

```ts
interface Math {
    max(...values: number[]): number
}

interface CallableFunction extends Function {
    apply<T, A extends any[], R>(this: (this: T, ...args: A) => R, thisArg: T, args: A): R
}
```
These signatures should be replaced by:

```ts
interface Math {
    max(...values: readonly number[]): number
}

interface CallableFunction extends Function {
    apply<T, A extends readonly any[], R>(this: (this: T, ...args: A) => R, thisArg: T, args: A): R
}
```
[Playground link](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gNyhNAXPHFMKhAJ7woCuAtgEYgwDaAugLABQAslBgAsAdHSgAPIVAAOUygApaECABoceNAEpu3AJYoMzAGZQwCPoPgBvbvFvxRYuUOe4INEIWIhS5KrUbM7BpE-kww3AC+2lx6BjDGpvAAwnjQDBAgAGI0KGAYOqjwIGIGKMBo8Nm5+YXWXHbw0rIUADwAKqoAgkUlIGUVJGQolI0oFOyqAEoAfHKCOp5zAgtEHfDOkjAA5p6dGvAAvNPwk6rzaJ3bq6qwO0R7RJOR3EA)

Basically every array rest parameter should be readonly, and there is other places in the library where readonly should be added. I already made these changes on a [fork](https://github.com/Conaclos/TypeScript-1/) (see the last commit of the fork). I'm not sure how to proceed, maybe the changes should be spitted in several ones?

Note that a lot of changes could be avoided if TypeScript implicitly inserted the readonly modifier for array rest parameter. I do not see any value to have mutable array rest parameter. However, this adds a special type case, and I am personally not a big fan of that.

**Comment by Jamesernator** [comment]
> However, this adds a special case in the type-checker, and I am personally not a big fan of that.

I wouldn't neccessarily call this a "special case". Rest parameters is a syntactical construct, plenty of other syntatical constructs change the types of things.

In fact this could even be done so that the internal type of the variable is unchanged, but the function rest parameter is wrapped with `Readonly`. i.e.:

```ts
function f(...args: number[]) {
    args.push(10); // fine
}

f([1,2,3] as ReadonlyArray<number>); // fine as f actually has type f(...readonly number[])
```

**Comment by DanielRosenwasser** [comment]
I don't think there's a need for any rest parameters to be rewritten (such as in `Math.max`). At most, you need to rewrite any *captured* argument list types `A` as `readonly [...A]`, right? So we could switch the type of `args`

```diff
interface CallableFunction extends Function {
-    apply<T, A extends any[], R>(this: (this: T, ...args: A) => R, thisArg: T, args: A): R
+    apply<T, A extends any[], R>(this: (this: T, ...args: A) => R, thisArg: T, args: readonly [...A]): R
}
```

**Comment by fatcerberus** [comment]
This is already allowed:
```ts
declare const vals: readonly number[]
Math.max(...vals);
```

Which makes sense because spreading an array always makes a copy.  So the problem is specific to `func.apply()`.

**Comment by Conaclos** [comment]
> I don't think there's a need for any rest parameters to be rewritten (such as in `Math.max`). At most, you need to rewrite any _captured_ argument list types `A` as `readonly [...A]`, right? So we could switch the type of `args`
> 
> ```diff
> interface CallableFunction extends Function {
> -    apply<T, A extends any[], R>(this: (this: T, ...args: A) => R, thisArg: T, args: A): R
> +    apply<T, A extends any[], R>(this: (this: T, ...args: A) => R, thisArg: T, args: readonly [...A]): R
> }
> ```

This is definitively a better proposal.

**Comment by Rudxain** [comment]
Has this issue been resolved?
